src/QsCompiler/Compiler/RewriteSteps/Monomorphization.cs

```diff
@@ -34,17 +36,19 @@ internal class Monomorphization : IRewriteStep
         /// Constructor for the Monomorphization Rewrite Step.
         /// </summary>
         /// <param name="monomorphizeIntrinsics">When true, intrinsics will be monomorphized as part of the rewrite step.</param>
-        public Monomorphization(bool monomorphizeIntrinsics = false)
+        /// <param name="isLibrary">When true, each public, non-generic callable will be treated as an entry point in the call graph.</param>
+        public Monomorphization(bool monomorphizeIntrinsics = false, bool isLibrary = false)
         {
             this.monomorphizeIntrinsics = monomorphizeIntrinsics;
             this.AssemblyConstants = new Dictionary<string, string?>();
+            this.isLibrary = isLibrary;
         }
 
-        public bool PreconditionVerification(QsCompilation compilation) => compilation.EntryPoints.Any();
+        public bool PreconditionVerification(QsCompilation compilation) => compilation.EntryPoints.Any() || this.isLibrary;
 
         public bool Transformation(QsCompilation compilation, out QsCompilation transformed)
         {
-            transformed = Monomorphize.Apply(compilation, this.monomorphizeIntrinsics);
+            transformed = Monomorphize.Apply(new QsCompilation(compilation.Namespaces, this.AugmentedEntryPoints(compilation)), this.monomorphizeIntrinsics);
```

Comment:
This would actually change all relevant callables to entry points for all transformation here on out, beyond just for monomorphization. Is that a good idea? Maybe we could capture the true entry points beforehand, do what you do here, but then remember to set the entry points field back to its original value before returning the tree.

Reply:
If the compiled project is an exe project, this will behave exactly as before (the original entry points just get passed through). For libraries, there are no original entrypoints (using that attribute in a library will cause a compile error). That said, you bring up a good point, in that right now the steps after this either don't care about entrypoints or know to treat them differently if it is a library. Any other rewrite steps could have the chance of getting confused by the compilation being updated this way. Originally I had considered updating the `QsCompilation` object to include a new field, but that would be a breaking change to the compiler.  I'm open to any thoughts on how to address this without affecting behavior for everyone. At least for now, any rewrite step that is active during QIR generation would believe that non-generic, non-internal callables in a library are actually entrypoints. That pretty much matches my intention, but I don't want to discount the possible side effects of this change.

Reply:
Since the EntryPoints fields is just a list of names of the entry point callables, it is fairly lightweight. My thinking would be to just capture the value of the EntryPoints field in a temporary variable before calling Monomorphize.Apply. Then when you do call the transformation, you can do the same logic you did here already by updating the tree so that all the callables are listed as entry points. The return from the transformation will be the monomorphized tree with all the relevant callables as entry points. Lastly you can easily replace the EntryPoints field in that transformed tree with the captured value of the field from before you called Monomorphization.Apply before returning the transformed tree. You are guaranteed that all the original entry points will still be in the tree, so you wouldn't have to worry about listing a callable that is not present anymore.

Reply:
That's a really good idea, thanks for suggesting! I can implement that approach and confirm it works for the target scenario.

Reply:
This actually helped me find a bug: the QirGeneration rewrite step also invokes SyntaxTreeTrimming but was not passing the new `isLibrary` bool to ensure the tree is trimmed based on non-internal, non-generic callables instead of solely entry points. I was also able to remove a check later in QirGeneration that was avoiding the incorrect entry points leftover from the Monomorphization. Thanks again!

Reply:
Glad it worked out 😁!